### ISSUE TITLE
Override single file extract location per app

### ISSUE THREAD

**Issue by IIFE** [dotnet/runtime#35249]
Hi,

Is it possible to specify a custom extract location for a single file exe without using `DOTNET_BUNDLE_EXTRACT_BASE_DIR`, or somehow apply `DOTNET_BUNDLE_EXTRACT_BASE_DIR` only to a specific exe (not through cmd however)?

Essentially I am building a self-contained single file app that runs as a Windows Service. The app can be installed by multiple consumers. If two consumers install the same single file exe, then the same temp location is used for the two Windows Service instances that get installed. I don't want that to happen as that causes uninstallation issues when one service is uninstalled but not another.

It would be useful if I can give each installation of the single file exe a unique location to extract to.

I can't set `DOTNET_BUNDLE_EXTRACT_BASE_DIR` as an environment variable because that will affect the whole system.

**Comment by IIFE** [comment]
@swaroop-sridhar

**Comment by IIFE** [comment]
Or potentially another option that ensures the single file exe is extracted to a new location if the existing temp location is being used by another single file exe?

**Comment by IIFE** [comment]
@swaroop-sridhar I think I may have found a workaround for the issue I'm facing, but would like to double check with you if there are any unwanted side effects.

The workaround is when each consumer installs the single file exe, it installs it with a unique name. So for example, if I produce a single file exe called `MySingleFile.exe`, consumer A will install it as `MySingleFile_v1.exe` and consumer B will install it as `MySingleFile_v2.exe`. Given that the extraction logic uses the bundle name to create the extraction directory, this workaround will ensure that each invocation of the exe will extract to a unique location under temp.

Does this workaround pose any issues? Or is it OK to rename a bundle file without it causing any side effects, apart from the extraction directory?

**Comment by IIFE** [comment]
@swaroop-sridhar Was wondering if it's possible to get any advice on the workaround I mentioned above? Seems to work, but I want to be sure there are not side effects.

**Comment by swaroop-sridhar** [comment]
Yes work-around will be OK, but I think it is better to set `DOTNET_BUNDLE_EXTRACT_BASE_DIR`  to a user-specific directory, ex: somewhere within %HOMEPATH%

**Comment by IIFE** [comment]
@swaroop-sridhar Unfortunately cannot rely on `DOTNET_BUNDLE_EXTRACT_BASE_DIR` because that affects the whole system. It's a desktop environment where the Windows service runs and not under my control. Is there no possible way to override the extract dir per app?

I think I read that in .NET 5 there will be no extraction for single file and it will all be mapped in memory any, so the issue should disappear with .NET 5?

**Comment by swaroop-sridhar** [comment]
> I think I read that in .NET 5 there will be no extraction for single file and it will all be mapped in memory any, so the issue should disappear with .NET 5?

Yes, I think most apps should be able to run without extraction in .net5. 
For native DLL dependencies, apps will have the option to leave the DLLs unbundled beside the app (default) or include them in the bundled (and extract at runtime). If some apps need to use the extraction option, and still need per-app extraction location, I think the best possible alternative is to implement https://github.com/dotnet/runtime/issues/3844.

**Comment by tapika** [comment]
There is a warp-tool mentioned in #60936 which can perform packing of folder with executable, where target folder is hardcoded within tool itself - in case if you want to use it immediately.

**Comment by elinor-fung** [comment]
Closing as we have moved away from extraction by default and have no plans to add a per-app configuration for extraction location.
